### PR TITLE
add keycard sign method

### LINKED ISSUE #7133: Sign transaction with Keycard
# Problem

It's not possible to sign transaction with Keycard

# Implementation

TBD

Screens https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Hardwallet-light-flows?node-id=0%3A1

# Acceptance Criteria

It should be possible to sign wallet transaction with Keycard 

screenshot 2018-12-18 17 05 11

<blockquote><div> Figma</div><div><strong><a href="https://www.figma.com/file/UfQjpWl1hmRchHIyY3Wvu2nW/Hardwallet-light-flows">Figma</a></strong></div><div>Created with Figma</div></blockquote>

### PR DESCRIPTION
**Pull request by dmitryn**
fixes #7133

### Summary:

Add ability to sign transactions and messages (for DApps) with keycard.

#### Platforms (optional)
- Android

status: ready

### CODE REVIEW

**Review by goranjovic — APPROVED** [review]

**Review by flexsurfer — APPROVED** [review]

**Comment by asemiankevich** [comment]
@dmitryn how do we want to handle [Sign message] case? Can't authorize myself in cryptokitties for example with keycard now.
Also, we need to deal with those 3 words since i've never seen them if account was created with keycard.


image

**Comment by asemiankevich** [comment]
@dmitryn 

1. seems PIN attempts are not renewed when enter wrong PIN and then a correct one.
image

2. PIN validation is broken on Sign transaction screen. If i enter wrong PIN when sending transaction - Enter New PIN screen appears without [Next] button and nothing happens (on android 9 Anna has seen endless spinner)

image

3. If sending transaction from nightly to this PR in 1x1 chat (open 1x1 chat -> send command) -> transaction is sent to wrong address. NOTE: if i send in wallet -> all is good and money are received

4. if card is upaired -> then i can't pair it to the account i have created in Status -> Error shown that card is not paired to this device and PIN screen is shown however I can't proceed any further.
If i want to pair with new account -> then Invalid card cryptogram is shown

image

5. also if i recover account with keycard here in the PR and same account is recovered in nightly - there is a difference in the 3 words name and default name so i can't sync the accounts.

**Comment by dmitryn** [comment]
@asemiankevich 
I've recently added support for signing messages for DApps based on https://github.com/status-im/status-go/pull/1421 that i initially missed to add (i've tested that by signing into cryptokitties DApp)

Next i'm going to address issues you have found above.

**Comment by asemiankevich** [comment]
@dmitryn seems some error is present in the code, builds can't be created

```
10:06:29  Retrieving org/mozilla/rhino/1.7R5/rhino-1.7R5.jar from central
10:06:29  Retrieving org/clojure/tools.analyzer.jvm/0.6.10/tools.analyzer.jvm-0.6.10.jar from central
10:06:29  Retrieving org/clojure/tools.analyzer/0.6.9/tools.analyzer-0.6.9.jar from central
10:06:29  Retrieving org/clojure/core.memoize/0.5.9/core.memoize-0.5.9.jar from central
10:06:29  Retrieving org/clojure/core.cache/0.6.5/core.cache-0.6.5.jar from central
10:06:29  Retrieving org/clojure/data.priority-map/0.0.7/data.priority-map-0.0.7.jar from central
10:06:29  Retrieving org/ow2/asm/asm-all/4.2/asm-all-4.2.jar from central
10:06:29  Retrieving org/clojure/tools.logging/0.3.1/tools.logging-0.3.1.jar from central
10:06:29  Retrieving com/yahoo/platform/yui/yuicompressor/2.4.8/yuicompressor-2.4.8.jar from central
10:06:29  Retrieving com/cognitect/transit-cljs/0.8.239/transit-cljs-0.8.239.jar from central
10:06:29  Retrieving com/cognitect/transit-js/0.8.846/transit-js-0.8.846.jar from central
10:06:29  Retrieving org/clojure/tools.reader/1.0.0-alpha4/tools.reader-1.0.0-alpha4.jar from central
10:06:29  Retrieving com/googlecode/java-diff-utils/diffutils/1.2.1/diffutils-1.2.1.jar from central
10:06:30  Retrieving lein-cljsbuild/lein-cljsbuild/1.1.7/lein-cljsbuild-1.1.7.jar from clojars
10:06:30  Retrieving lein-re-frisk/lein-re-frisk/0.5.8/lein-re-frisk-0.5.8.jar from clojars
10:06:30  Retrieving re-frisk-sidecar/re-frisk-sidecar/0.5.7/re-frisk-sidecar-0.5.7.jar from clojars
10:06:30  Retrieving fs/fs/1.1.2/fs-1.1.2.jar from clojars
10:06:30  Retrieving ring-cors/ring-cors/0.1.8/ring-cors-0.1.8.jar from clojars
10:06:31  Retrieving ring/ring-core/1.4.0/ring-core-1.4.0.jar from clojars
10:06:31  Retrieving crypto-random/crypto-random/1.2.0/crypto-random-1.2.0.jar from clojars
10:06:31  Retrieving crypto-equality/crypto-equality/1.0.0/crypto-equality-1.0.0.jar from clojars
10:06:31  Retrieving ring/ring-defaults/0.1.5/ring-defaults-0.1.5.jar from clojars
10:06:31  Retrieving clj-time/clj-time/0.9.0/clj-time-0.9.0.jar from clojars
10:06:31  Retrieving ring/ring-ssl/0.2.1/ring-ssl-0.2.1.jar from clojars
10:06:32  Retrieving ring/ring-anti-forgery/1.0.0/ring-anti-forgery-1.0.0.jar from clojars
10:06:32  Retrieving ring/ring-headers/0.1.3/ring-headers-0.1.3.jar from clojars
10:06:32  Retrieving http-kit/http-kit/2.2.0/http-kit-2.2.0.jar from clojars
10:06:32  Retrieving hiccup/hiccup/1.0.5/hiccup-1.0.5.jar from clojars
10:06:32  Retrieving com/taoensso/encore/2.84.2/encore-2.84.2.jar from clojars
10:06:32  Retrieving com/taoensso/truss/1.3.6/truss-1.3.6.jar from clojars
10:06:32  Retrieving io/aviso/pretty/0.1.30/pretty-0.1.30.jar from clojars
10:06:32  Retrieving com/taoensso/sente/1.11.0/sente-1.11.0.jar from clojars
10:06:33  Retrieving com/taoensso/timbre/4.7.4/timbre-4.7.4.jar from clojars
10:06:33  Retrieving clout/clout/2.1.2/clout-2.1.2.jar from clojars
10:06:33  Retrieving ring/ring-codec/1.0.1/ring-codec-1.0.1.jar from clojars
10:06:33  Retrieving compojure/compojure/1.5.2/compojure-1.5.2.jar from clojars
10:06:34  Retrieving medley/medley/0.8.2/medley-0.8.2.jar from clojars
10:06:34  Retrieving instaparse/instaparse/1.4.0/instaparse-1.4.0.jar from clojars
10:06:34  Retrieving reagent/reagent/0.7.0/reagent-0.7.0.jar from clojars
10:06:34  Retrieving cljsjs/react-dom/15.5.4-0/react-dom-15.5.4-0.jar from clojars
10:06:34  Retrieving cljsjs/create-react-class/15.5.3-0/create-react-class-15.5.3-0.jar from clojars
10:06:34  Retrieving cljsjs/react/15.5.4-0/react-15.5.4-0.jar from clojars
10:06:35  Retrieving re-frame/re-frame/0.10.1/re-frame-0.10.1.jar from clojars
10:06:35  Retrieving net/cgrand/macrovich/0.2.0/macrovich-0.2.0.jar from clojars
10:06:35  Retrieving cljsjs/react-dom-server/15.5.4-0/react-dom-server-15.5.4-0.jar from clojars
10:06:35  Retrieving re-frisk-shell/re-frisk-shell/0.5.2/re-frisk-shell-0.5.2.jar from clojars
10:06:35  Retrieving com/andrewmcveigh/cljs-time/0.4.0/cljs-time-0.4.0.jar from clojars
10:06:36  Retrieving re-frisk/re-frisk/0.5.4/re-frisk-0.5.4.jar from clojars
10:06:36  Retrieving re-com/re-com/2.1.0/re-com-2.1.0.jar from clojars
10:06:36  Retrieving day8/re-frame/re-frame-10x/0.3.2/re-frame-10x-0.3.2.jar from clojars
10:06:36  Retrieving binaryage/devtools/0.9.10/devtools-0.9.10.jar from clojars
10:06:36  Retrieving binaryage/env-config/0.2.2/env-config-0.2.2.jar from clojars
10:06:36  Retrieving zprint/zprint/0.4.7/zprint-0.4.7.jar from clojars
10:06:36  Retrieving cljsjs/react-flip-move/2.9.17-0/react-flip-move-2.9.17-0.jar from clojars
10:06:37  Retrieving cljsjs/prop-types/15.6.0-0/prop-types-15.6.0-0.jar from clojars
10:06:37  Retrieving cljsjs/react-highlight/1.0.7-1/react-highlight-1.0.7-1.jar from clojars
10:06:37  Retrieving lein-cljfmt/lein-cljfmt/0.5.7/lein-cljfmt-0.5.7.jar from clojars
10:06:37  Retrieving cljsjs/highlight/9.6.0-0/highlight-9.6.0-0.jar from clojars
10:06:37  Retrieving rewrite-cljs/rewrite-cljs/0.4.3/rewrite-cljs-0.4.3.jar from clojars
10:06:37  Retrieving rewrite-clj/rewrite-clj/0.5.2/rewrite-clj-0.5.2.jar from clojars
10:06:38  Retrieving cljfmt/cljfmt/0.5.7/cljfmt-0.5.7.jar from clojars
10:06:38  Retrieving rasom/lein-githooks/0.1.5/lein-githooks-0.1.5.jar from clojars
10:06:38  Retrieving meta-merge/meta-merge/1.0.0/meta-merge-1.0.0.jar from clojars
10:06:39  Updating hook:  pre-commit
10:07:01  src/status_im/ui/screens/wallet/send/events.cljs has incorrect formatting
10:07:01  --- a/src/status_im/ui/screens/wallet/send/events.cljs
10:07:01  +++ b/src/status_im/ui/screens/wallet/send/events.cljs
10:07:01  [036m@@ -359,17 +359,17 @@[0m
10:07:01          {::hash-message {:message      (ethereum/naked-address data)
10:07:01                           :on-completed #(re-frame/dispatch [::hash-message-completed (types/json->clj %)])}}
10:07:01          (if typed?
10:07:01  [031m-              {::sign-typed-data {:data     data[0m
10:07:01  [031m-                                  :password password[0m
10:07:01  [031m-                                  :account  from[0m
10:07:01  [031m-                                  :on-completed #(re-frame/dispatch [::sign-message-completed[0m
10:07:01  [031m-                                                                     screen-params[0m
10:07:01  [031m-                                                                     (types/json->clj %)[0m
10:07:01  [031m-                                                                     password-error-cb])}}[0m
10:07:01  [031m-              {::sign-message {:params       {:data     data[0m
10:07:01  [031m-                                              :password (security/safe-unmask-data password)[0m
10:07:01  [031m-                                              :account  from}[0m
10:07:01  [031m-                               :on-completed #(re-frame/dispatch [::sign-message-completed[0m
10:07:01  [031m-                                                                  screen-params[0m
10:07:01  [031m-                                                                  (types/json->clj %)[0m
10:07:01  [031m-                                                                  password-error-cb])}})))))[0m
10:07:01  [032m+         {::sign-typed-data {:data     data[0m
10:07:01  [032m+                             :password password[0m
10:07:01  [032m+                             :account  from[0m
10:07:01  [032m+                             :on-completed #(re-frame/dispatch [::sign-message-completed[0m
10:07:01  [032m+                                                                screen-params[0m
10:07:01  [032m+                                                                (types/json->clj %)[0m
10:07:01  [032m+                                                                password-error-cb])}}[0m
10:07:01  [032m+         {::sign-message {:params       {:data     data[0m
10:07:01  [032m+                                         :password (security/safe-unmask-data password)[0m
10:07:01  [032m+                                         :account  from}[0m
10:07:01  [032m+                          :on-completed #(re-frame/dispatch [::sign-message-completed[0m
10:07:01  [032m+                                                             screen-params[0m
10:07:01  [032m+                                                             (types/json->clj %)[0m
10:07:01  [032m+                                                             password-error-cb])}})))))[0m
10:07:27  1 file(s) formatted incorrectly
```

**Comment by dmitryn** [comment]
@asemiankevich i've fixed code formatting, build is ready

http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190321-074649-a8c35b-pr.apk

**Comment by asemiankevich** [comment]
@dmitryn 
-  i can't bypass sign in in Cryptokitties, it redirects me to this screen over and over again. It works in develop.
- I can sign a message in Test Dapp on Ropsten, but can't send Typed sign message

image

- enter PIN screen is a bit broken

image

**Comment by dmitryn** [comment]
@asemiankevich signing typed messages is not ready yet, i'm waiting for status-go implementation to proceed with that. 

Regarding cryptokitties, i had redirect issue yesterday, but after i rebased onto develop it's gone. I'll check again.

**Comment by dmitryn** [comment]
@asemiankevich i've added signed typed data support. Please take a look.

Comments on bugs:

1. Fixed

2. Fixed

3. Sent tx on ropsten from nightly 1x1 to this PR and got funds. Cannot reproduce.

4.  Seems not related to tx/message signing. I suggest to open issues for those bugs separately. If tx/message signing work, so we would have all functionality implemented (in happy case at least). Then we will work on bugs.

5. Same as 4

**Comment by asemiankevich** [comment]
@dmitryn endless spinner when signing transaction with invalid PIN is still reproducible here. Do you want to fix negative cases in separate PRs btw? Not sure how hard is to keep the branch up to date. @guylouis what do you think?

**Comment by dmitryn** [comment]
@asemiankevich Endless spinner should be now fixed, please take a look http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190325-081457-b1fdfc-pr.apk

**Comment by asemiankevich** [comment]
@dmitryn can we have this Got it screen back once transaction is signed? It should be shown for all transactions no matter it is confirmed with password or keycard

image

**Comment by dmitryn** [comment]
@asemiankevich yes, i will add it for keycard as well.

**Comment by dmitryn** [comment]
@asemiankevich screen added, please take a look 

http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190325-161635-16339d-pr.apk

**Comment by asemiankevich** [comment]
@dmitryn is it possible that previous commit which fixed endless spinner is gone? It was working in previous build and in latest one i see this endless spinner when using wrong PIN again

**Comment by dmitryn** [comment]
@asemiankevich should be fixed now, please check

**Comment by asemiankevich** [comment]
@dmitryn if i sign `message` to dapp with invalid PIN -> user is not stayed on the screen where he should be able to re-enter PIN , instead the message is signed. However i see the screen when i am trying to sign the message again:

image


Steps:
- open status
- switch to ropsten
- open dapps -> test dapp
- tap transactions
- tap sign message to dapp
- enter invalid pin
Note that screen is closed and message is signed.
- tap the sign message again in the test dapp

Also, PUK code validation is broken, endless spinner when entering invalid PUK (after 3 invalid pin attempts on login screen)

**Comment by dmitryn** [comment]
Made the fixes, but now we need to wait for new status-go version to be released as status-react develop now uses ` 0.23.0-beta.8-chaos` which doesn't have `hashMessage` and `hashTypedData` methods so current build is failing. 

Waiting for https://github.com/status-im/status-go/pull/1426 to be reviewed and merged first.

**Comment by dmitryn** [comment]
@asemiankevich new build is finally ready http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190328-174621-0a9ae4-pr.apk

**Comment by asemiankevich** [comment]
yeah, on it already @dmitryn :)

**Comment by asemiankevich** [comment]
@dmitryn @guylouis @hesterbruikman do we need to show user `set up wallet screen` if account was set up with keycard? I assume those 3 words are replaced with keycard in this case, so there is no need for wallet set up here. But we need to leave it if user creates account with password (no keycard involved). 
Is it right?

1. open status
2. switch to ropsten
3. open Test dapp (dapps -> test dapp)
4. assets - request eth
5. transactions -> deploy simple contract

image

**Comment by asemiankevich** [comment]
also can't re-initiate transaction in Test Dapp if i changed my mind and tapped back being on enter pin screen:

- open status
- switch to ropsten
- open Test dapp (dapps -> test dapp)
- assets - request eth
- transactions -> request stt
- tap sign transaction
- tap Back button when being on Enter PIN screen
- tap Request STT again

@flexsurfer maybe there are some changes that we can do to the dapp itself? this scenario works fine in develop (without PIN request screen ofc)

**Comment by dmitryn** [comment]
@asemiankevich i've removed wallet onboarding (wallet setup screen) for keycard accounts.
Also, i've fixed issue with navigating back from Enter Pin screen and re-initiating tx from DApp. 

http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190328-193539-7b5f7a-pr.apk

**Comment by guylouis** [comment]
@asemiankevich @dmitryn 
The three words phrase is independant from where is stored the private key (ie keycard or the phone it self), it's here to prevent impersonating the wallet screens in the app. So we still need it with keycard accounts.

**Comment by asemiankevich** [comment]
> @asemiankevich @dmitryn
> The three words phrase is independant from where is stored the private key (ie keycard or the phone it self), it's here to prevent impersonating the wallet screens in the app. So we still need it with keycard accounts.

hold on. Do you want to have this 3-words-confirmation along with keycard?
If yes - then it is confusing since confirming that 3 wrords are correct requires password. And we dont have password, we have keycard ( if account was set up with keycard).

image

Please clarify more on that.

**Comment by guylouis** [comment]
What would make sense is (with keycard account) to have the user check his 3 words and confirm they are ok , but not input password again (he does not have it). And have PIN entered only once during the signing process. If this raises technical issues, let's discuss them.

**Comment by asemiankevich** [comment]
@dmitryn please evaluate the effort here. @hesterbruikman please review

**Comment by guylouis** [comment]
@goranjovic and @corpetty opinon needed for the security side here too

**Comment by dmitryn** [comment]
Screenshot_20190329-121127

This is the Send tx screen. Do we want to put that 3-words right  after "Advanced" button?

**Comment by hesterbruikman** [comment]
> 
> 
> @dmitryn please evaluate the effort here. @hesterbruikman please review

IMO requires Wallet setup as well as Signing phrase in transaction flow as I understood the contents of the transaction information can be misrepresented through spoofing. e.g. requesting higher amount than is displayed.

That leaves the issue that a password is requested that is not created when initializing Keycard. I think Signing phrase and Password need to be seperated in the screen regardless; my main concern with the current design. Looking into this now.

**Comment by guylouis** [comment]
@hesterbruikman wdyt ?
three possibilities:
- on this send transaction screen in the bottom (need some redesign)
- on the pin entry screen (need some redesign)
- in a new screen we put between the two :)

**Comment by hesterbruikman** [comment]
Overview with address@2x

For consistency I would go for the send transaction screen. This is not waterproof, but I think we need to put some more thought into signingphrase across the app that is not worth blocking this issue now as we're following the existing approach. 

@dmitryn please note there's a 10% opacity layer on top of the content below the bottom sheet to handle the tooltip.

**Comment by dmitryn** [comment]
@hesterbruikman we currently use another design for wallet Send screen 

Screenshot_20190329-125959

For now i just put 3-words under Advanced button, not sure there is enough space to show tooltip unless screen is scrollable

**Comment by hesterbruikman** [comment]
> For now i just put 3-words under Advanced button, not sure there is enough space to show tooltip unless screen is scrollable

When do you expect this being tested and merged? Can we use the updated screen if @goranjovic 's wallet redesign comes through?

As a backup, let me deliver you a mockup for the screen you're using now.

**Comment by dmitryn** [comment]
@hesterbruikman i'd like to merge this PR today while we have @asemiankevich here to help with testing.

As for new designs, yes, we can update that screen to new design after new wallet design comes through.

**Comment by dmitryn** [comment]
Screenshot_20190329-135520

Also, i've come up with this screen for now. Maybe we don't need to put much effort into current screen design if we will be using new design soon.


Have a call now but I'll get you an update within the hour.

**Comment by hesterbruikman** [comment]
Frame 2 5

Signing phrase is text input with height 40px (same as Advanced button) with 16px padding top and bottom. With a 5 spaces between words; not sure how to set word spacing.

Text below is caption: https://www.figma.com/file/cb4p8AxLtTF3q1L6JYDnKN15/Index?node-id=787%3A1082 
EDIT: text has a padding of 0 at the top and 14 at the bottom

**Comment by dmitryn** [comment]
@asemiankevich new build is ready http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190329-123740-0f486e-pr.apk

**Comment by asemiankevich** [comment]
> @asemiankevich new build is ready http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190329-123740-0f486e-pr.apk

on it! :)

**Comment by asemiankevich** [comment]
@dmitryn @hesterbruikman @guylouis i tested this one and suggest to merge as is. There is 1 issue we have now that could be fixed separately (missing translation at the bottom)
image

**Comment by dmitryn** [comment]
@asemiankevich i've just pushed commit with missing translations

**Comment by dmitryn** [comment]
New build http://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190329-144643-49f01d-pr.apk

File: STATUS_GO_VERSION [hunk 1]
```diff
@@ -1 +1 @@
-0.23.0-beta.8-chaos
+190328-172624-dc2531
```
Comment (dmitryn) [inline STATUS_GO_VERSION]: I'm putting this PR on hold until we will release new status-go version. 

@gravityblast do you think we can use this guide to make new release? https://github.com/status-im/status-go/blob/develop/RELEASING.md

Comment (dmitryn) [reply STATUS_GO_VERSION]: I guess we could just switch to `0.23.0-beta.10-chaos` release

Comment (gravityblast) [reply STATUS_GO_VERSION]: I'm not sure if the next release will continue from that branch or not, so I would wait for tomorrow for this

**Pull request merged**